cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/ArtifactCacheManager.java

```diff
@@ -0,0 +1,203 @@
+        if (instance == null) {
+          instance = new ArtifactCacheManager();
+        }
+      }
+    }
+    return instance;
+  }
+
+  /**
+   * Increases cache usage counter for all the files(cached artifacts) in the GCS bucket.
+   */
+  public void recordCacheUsageForArtifacts(Storage client, String bucket, Set<String> files, String runRootPath,
+                                           String cachePath) {
+    try {
+      String cachedArtifactsListFilePath = getPath(runRootPath, CACHED_ARTIFACTS_LIST_FOR_RUNID);
+      storeCachedArtifactsForRun(client, bucket, files, cachedArtifactsListFilePath);
+      String cacheCountFilePath = getPath(cachePath, CACHED_ARTIFACTS_USAGE_COUNT);
+      changeUsageCountForArtifacts(client, bucket, files, cachePath, cacheCountFilePath, 1);
+    } catch (Exception e) {
+      LOG.warn("Error in recordCacheUsageForArtifacts", e);
+    }
+  }
+
+  /**
+   * Decreases cache usage counter for all the files(cached artifacts) in the GCS bucket.
+   * Also updates custom_time and removes Temporary Hold if the artifact is not being used by any program run.
+   */
+  public void releaseCacheUsageForArtifacts(Storage client, String bucket, String runRootPath) {
+    try {
+      String cachedArtifactsListFilePath = getPath(runRootPath, CACHED_ARTIFACTS_LIST_FOR_RUNID);
+      Set<String> files = getCachedArtifactsForRun(client, bucket, cachedArtifactsListFilePath);
+      String cachedArtifactsPath = getPath(DataprocUtils.CDAP_GCS_ROOT, DataprocUtils.CDAP_CACHED_ARTIFACTS);
+      String cacheCountFilePath = getPath(cachedArtifactsPath, CACHED_ARTIFACTS_USAGE_COUNT);
+      changeUsageCountForArtifacts(client, bucket, files, cachedArtifactsPath, cacheCountFilePath, -1);
+    } catch (Exception e) {
+      LOG.warn("Error in releaseCacheUsageForArtifacts", e);
+    }
+  }
+
+  private void changeUsageCountForArtifacts(Storage client, String bucket, Set<String> files, String cachePath,
+                                            String filePath, int changeValue) throws InterruptedException {
+    if (files.isEmpty()) {
+      LOG.debug("No Cached Artifacts found for this run!");
+      return;
+    }
+    BlobId blobId = BlobId.of(bucket, filePath);
+    for (int i = 0; i < MAX_RETRIES_FOR_CACHE_COUNTER_OPERATION; i++) {
+      try {
+        Blob blob = client.get(blobId);
+        if (blob == null || !blob.exists()) {
+          LOG.debug("Creating ArtifactsCacheCounter {} at {}", CACHED_ARTIFACTS_USAGE_COUNT, filePath);
+          blob = createCacheCounterFile(client, blobId);
+        }
+        Map<String, Integer> artifactCount = GSON.fromJson(new String(blob.getContent(), StandardCharsets.UTF_8),
+                                                           new TypeToken<Map<String, Integer>>() {
+                                                           }.getType());
+        modifyCacheCounter(client, bucket, files, cachePath, artifactCount, changeValue);
```

Comment:
Why we need to embed `setCustomTimeOnArtifactAndReleaseHold` in `modifyCacheCounter` is confusing?
Why not calling it similar to `writeCacheCounterToGCS` ?

Reply:
Done